src/main/java/com/glencoesoftware/omero/ms/image/region/HistogramRequestHandler.java

```diff
@@ -0,0 +1,300 @@
+
+    private static final org.slf4j.Logger log = LoggerFactory
+            .getLogger(HistogramRequestHandler.class);
+
+    /** Histogram Request Context */
+    HistogramCtx histogramCtx;
+
+    /** OMERO server pixels service. */
+    private PixelsService pixelsService;
+
+    /**
+     * Constructor. Populates histogramCtx and pixelsService
+     * @param histogramCtx
+     * @param pixelsService
+     */
+    public HistogramRequestHandler(HistogramCtx histogramCtx,
+            PixelsService pixelsService) {
+        this.histogramCtx = histogramCtx;
+        this.pixelsService = pixelsService;
+    }
+
+    /**
+     * Get the min and max for the channel from the StatsInfo
+     * @param channel The channel to get the StatsInfo min and max for
+     * @return double array [min, max] if the StatsInfo is present, else null
+     */
+    private double[] getMinMaxFromStatsinfo(Channel channel) {
+        double min, max;
+        if (channel.getStatsInfo() != null) {
+            min = channel.getStatsInfo().getGlobalMin();
+            max = channel.getStatsInfo().getGlobalMax();
+            // if max == 1.0 the global min/max probably has not been
+            // calculated; fall back to plane min/max
+            if (max != 1.0)
+                return new double[] { min, max };
+        }
+        return null;
+    }
+
+    /**
+     * Get the minimum and maximum value of the pixel data
+     *
+     * @param pd
+     *            The {@link PixelData}
+     * @param channel
+     *            The {@link Channel}
+     * @return See above
+     */
+    private double[] getMinMaxFromPixelData(PixelData pd, Channel channel) {
+        double min, max;
+
+        StatsFactory sf = new StatsFactory();
+        double[] pixelMinMax = sf.initPixelsRange(channel.getPixels());
+
+        min = pixelMinMax[1];
+        max = pixelMinMax[0];
+
+        for (int i = 0; i < pd.size(); i++) {
+            min = Math.min(min, pd.getPixelValue(i));
+            max = Math.max(max, pd.getPixelValue(i));
+        }
+
+        return new double[] { min, max };
+    }
+
+    /**
+     * Read through the pixel data and produce histogram data reflecting
+     * the occurrance of pixel values within each bin
+     * @param pd The {@link PixelData} to get the pixel values from
+     * @param channel The {@link Channel} to get the pixel values from
+     * @param minMax The min and max to values to divide into bins
+     * @param imgWidth The sizeX of the pixel data
+     * (not necessarily full resolution)
+     * @param imgHeight The sizeY of the pixel data
+     * (not necessarily full resolution)
+     * @return {@link JsonArray} containing histogram data
+     */
+    public JsonArray getHistogramData(PixelData pd,
+            double[] minMax, int imgWidth, int imgHeight) {
+        int[] counts = new int[histogramCtx.bins];
+
+        double min = minMax[0];
+        double max = minMax[1];
+
+        double range = max - min + 1;
```

Comment:
I might be missing something obvious but the increment by one is puzzling me. Taking the example of floating point data with real values between 0.3 and 0.5, this would lead to a range of 1.2 which feels incorrect.

It looks like this was introduced server-side in https://github.com/ome/omero-server/commit/2037f7a87e10ee50b4c23e16abe3024e9904921f but I don't see much more context in the commit message or the original PR (https://github.com/ome/openmicroscopy/pull/4931) about the scenario this is meant to address.

Reply:
Discussed quickly with Kevin, I think we are both in agreement that there are shortcomings with the current bin assignment strategy in `omero-server` and mirrored in this PR.

An issue with trying to fix the implementation as part of this PR is the added complexity in terms of testing since the response of the micro-service will not match the response of the server.

To simplify the review process, I agree with Kevin's suggestion to restrict the scope of this PR to the addition of the histogram endpoint. We can then open an upstream issue discussing the histogram calculation logic, with concrete use cases and scenarios, and propose changes that would be applicable to the `omero-server` component as well as the micro-service and ideally should be released in locksteps